src/parser/parser.cpp

```diff
@@ -138,39 +137,78 @@ void Parser::ParseQuery(const string &query) {
 	}
 	{
 		PostgresParser::SetPreserveIdentifierCase(options.preserve_identifier_case);
-		PostgresParser parser;
-		parser.Parse(query);
-		if (parser.success) {
-			if (!parser.parse_tree) {
-				// empty statement
-				return;
-			}
+		bool parsing_failed = false;
+		// Creating a new scope to prevent multiple PostgresParser destructors being called
+		// which led to some memory issues
+		{
+			PostgresParser parser;
+			parser.Parse(query);
+			if (parser.success) {
+				if (!parser.parse_tree) {
+					// empty statement
+					return;
+				}
 
-			// if it succeeded, we transform the Postgres parse tree into a list of
-			// SQLStatements
-			transformer.TransformParseTree(parser.parse_tree, statements);
-		} else {
-			parser_error = QueryErrorContext::Format(query, parser.error_message, parser.error_location - 1);
+				// if it succeeded, we transform the Postgres parse tree into a list of
+				// SQLStatements
+				transformer.TransformParseTree(parser.parse_tree, statements);
+			} else {
+				parsing_failed = true;
+			}
 		}
-	}
-	if (!parser_error.empty()) {
-		if (options.extensions) {
-			for (auto &ext : *options.extensions) {
-				D_ASSERT(ext.parse_function);
-				auto result = ext.parse_function(ext.parser_info.get(), query);
-				if (result.type == ParserExtensionResultType::PARSE_SUCCESSFUL) {
-					auto statement = make_uniq<ExtensionStatement>(ext, std::move(result.parse_data));
-					statement->stmt_length = query.size();
-					statement->stmt_location = 0;
-					statements.push_back(std::move(statement));
-					return;
+		// if DuckDB fails to parse the entire sql string, break the string down into individual statements
+		// so that parser extensions can help parse some statements
+		if (parsing_failed) {
+			vector<string> query_statements;
+			auto tokens = Tokenize(query);
+			auto next_statement_start = 0;
+			for (size_t i = 1; i < tokens.size(); ++i) {
+				auto &t_prev = tokens[i - 1];
+				auto &t = tokens[i];
+				if (t_prev.type == SimplifiedTokenType::SIMPLIFIED_TOKEN_OPERATOR) {
+					if (StringUtil::Contains(query.substr(t_prev.start, t.start - t_prev.start), ";")) {
```

Comment:
This makes a copy of the substring for every token - a straightforward loop (as we are looking for a single character) would be more performant

Reply:
Done